### ISSUE TITLE
[FEATURE]: use git SHA in remote source declaration

### ISSUE THREAD

**Issue by hnicke** [garden-io/garden#3172]
### Background / Motivation

We are running a poly repo setup.
The `project.garden.yml` lives in an umbrella repository which declares all other repos as remote sources.
In our CI pipeline of the individual repos we trigger a child pipeline in the umbrella repo where `garden test && garden deploy` is run.
When triggering this child pipeline we pass a specific context as payload to the umbrella pipeline:
- which repo has triggered this pipeline run
- from which git SHA was this pipeline triggered

The idea is to integrate the specific git commit which triggered the pipeline with the other repos main development branches.

### What should the user be able to do?
We'd like to use garden templating to adjust the remote source definition of the project.
In the remote source declaration we would like to be able to point to a git SHA.
Currently, only branches are allowed.

### Why do they want to do this? What problem does it solve?

This way we could use gardens templating capabilities to use a specific git SHA instead of the main development branch in our CI pipeline.

Our current workaround is in our pipeline to clone the specific repository manually, checkout the required git SHA and use `garden link`.

### How important is this feature for you/your team?

🌹 It’s a nice to have, but nice things are nice

**Comment by Orzelius** [comment]
The reason not to do it originally was that we couldn't do shallow clones from GitHub with commit SHAs. This may have changed since.
It makes sense to support this and just fallback on a full clone. We'll look into implementing. Thank you for bringing this to our attention and sorry for the late reply.

**Comment by hnicke** [comment]
Still relevant.

**Comment by Orzelius** [comment]
Hello @hnicke, this has noow been implemented via #3498. Thank you for bringing this to our attention. There's a chance the submodules will misbehave if a direct commit is used. If that happens we would love to look further into that and would appreciate a repro.

**Comment by hnicke** [comment]
Thank you! This will simplify our CI flow.

We'll definitely try it out once it's released and report any errors.

**Comment by hnicke** [comment]
After #3873 has been implemented I was able to test #3172.
It seems to work well with long git SHAs.
Unfortunately, short SHAs don't work:
```
Downloading remote project failed with error: 
Error: Command "git -c protocol.file.allow=always clone --recursive --depth=1 --shallow-submodules --branch=442cb8443f4c git@bitbucket.org:xxxx/xxxxx.git ." failed with code 128:
Cloning into '.'...
Warning: Permanently added the ECDSA host key for IP address 'xxxxxxxxx' to the list of known hosts.
warning: Could not find remote branch 442cb8443f4c to clone.
fatal: Remote branch 442cb8443f4c not found in upstream origin
fatal: the remote end hung up unexpectedly
```

Would it be possible to support short SHAs as well?
I am aware that it's possible to derive the long SHA from the short SHA by running `git rev-parse $SHORT_SHA` - however this requires the repo to be checked out manually.

**Comment by Walther** [comment]
@hnicke Unfortunately, the shallow clone for a specific commit requires a full 40char hash:
> `<refs>`...
> The remote heads to update from. This is relative to $GIT_DIR (e.g. "HEAD", "refs/heads/master"). When unspecified, update from all heads the remote side has.
> 
> If the remote has enabled the options uploadpack.allowTipSHA1InWant, uploadpack.allowReachableSHA1InWant, or uploadpack.allowAnySHA1InWant, they may alternatively be 40-hex sha1s present on the remote.

https://git-scm.com/docs/git-fetch-pack#Documentation/git-fetch-pack.txt-ltrefsgt82308203

> Starting with Git 2.5.0 (which needs to be available at both the client and server side) you can set uploadpack.allowReachableSHA1InWant=true on the server side to enable fetching of specific SHA1s (must be the full SHA1, not an abbreviation)

https://stackoverflow.com/a/43136160
